### ISSUE TITLE
How to handle firebase auth errors?

### ISSUE THREAD

**Issue by kroikie** [FirebaseExtended/flutterfire#725]
I'm trying to build a login page with firebase auth plugin. But I don't know how to handle auth errors in a cross-platform way. For example, this are some `PlatformException` that I get when the user type the wrong password/email or there are a network issue.

On iOS:

```
PlatformException(Error 17011, FIRAuthErrorDomain, There is no user record corresponding to this identifier. The user may have been deleted.)

PlatformException(Error 17009, FIRAuthErrorDomain, The password is invalid or the user does not have a password.)

PlatformException(Error 17020, FIRAuthErrorDomain, Network error (such as timeout, interrupted connection or unreachable host) has occurred.)
```

On Android:

```
PlatformException(exception, There is no user record corresponding to this identifier. The user may have been deleted., null)

PlatformException(exception, The password is invalid or the user does not have a password., null)

PlatformException(exception, A network error (such as timeout, interrupted connection or unreachable host) has occurred., null)
```

The format is: code, message, details. In Android I get a code: `exception`, what does it mean? And also in iOS I always get a message: `FIRAuthErrorDomain` and not any standard _code_ listed [here](https://firebase.google.com/docs/auth/ios/errors).

The code:

```dart
try {
    _user = await _auth.signInWithEmailAndPassword(
        email: _email,
        password: _password);
} on PlatformException catch (e) {
    print(e);
}
```

## Temporary fix (don't rely on this solution)

In the meantime the problem can be fixed  (see bottom for warning on this solution) with something like that:

```dart
enum authProblems { UserNotFound, PasswordNotValid, NetworkError }
```
```dart
authProblems errorType;
if (Platform.isAndroid) {
  switch (e.message) {
    case 'There is no user record corresponding to this identifier. The user may have been deleted.':
      errorType = authProblems.UserNotFound;
      break;
    case 'The password is invalid or the user does not have a password.':
      errorType = authProblems.PasswordNotValid;
      break;
    case 'A network error (such as timeout, interrupted connection or unreachable host) has occurred.':
      errorType = authProblems.NetworkError;
      break;
    // ...
    default:
      print('Case ${e.message} is not jet implemented');
  }
} else if (Platform.isIOS) {
  switch (e.code) {
    case 'Error 17011':
      errorType = authProblems.UserNotFound;
      break;
    case 'Error 17009':
      errorType = authProblems.PasswordNotValid;
      break;
    case 'Error 17020':
      errorType = authProblems.NetworkError;
      break;
    // ...
    default:
      print('Case ${e.message} is not jet implemented');
  }
}
print('The error is $errorType');
```

**Warning**: The problem, that I would like to emphasize with this issue, it's that you can't rely too much in this solution because a simple FirebaseAuth update can screw it up. Also if you choose this way be sure to lock the `firebase_auth: ?.?.?` to a specific version in the pubspec.yaml file.

----

Flutter doctor

```
[✓] Flutter (Channel beta, v0.5.1, on Mac OS X 10.13.6 17G65, locale en-IT)
    • Flutter version 0.5.1 at /Users/.../Library/flutter
    • Framework revision c7ea3ca377 (10 weeks ago), 2018-05-29 21:07:33 +0200
    • Engine revision 1ed25ca7b7
    • Dart version 2.0.0-dev.58.0.flutter-f981f09760

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/.../Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] IntelliJ IDEA Community Edition (version 2017.3.5)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 25.0.1
    • Dart plugin version 173.4700

[!] VS Code (version 1.25.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected devices (1 available)
    • Nexus 5X • 01ce0d68fc7ca517 • android-arm64 • Android 8.1.0 (API 27)
```

**Comment by Yazazzello** [comment]
I came here from 'How to handle firebase auth errors? #20223' where @kroikie said that any further  collaboration will be done here. And both issues are closed without a solution? Is that right or I missed something?

**Comment by JonasTriki** [comment]
Any update on this? Why is it closed?

**Comment by Dararii** [comment]
I think we don't have any solution for this right now. sad.

**Comment by getgoh** [comment]
Here's the file we are all looking for. Error messages are listed in comments on top of each method.

https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart

**Comment by mhmoudsabry** [comment]
> Here's the file we are all looking for. Error messages are listed in comments on top of each method.
> 
> https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart

i think this is the solution i need thanks

**Comment by takuro-kira** [comment]
Error list on each method is good!
I can handle exception like this.

```
    final FirebaseUser user = (await _auth.createUserWithEmailAndPassword(
      email: email,
      password: password,
    ).catchError((error) {
      switch (error.code) {
        case "ERROR_WEAK_PASSWORD":
          break;
        case "ERROR_INVALID_EMAIL":
          break;
        case "ERROR_EMAIL_ALREADY_IN_USE":
          break;
        default:
          print(error);
          break;
      }
    })).user;
```

But is there any way cover code typo?

**Comment by saqibsoftwares** [comment]
> Here's the file we are all looking for. Error messages are listed in comments on top of each method.
> 
> https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart

Still missing some codes, like "ERROR_INVALID_VERIFICATION_CODE"

**Comment by feliwir** [comment]
Any updates on how to handle these errors?

**Comment by AlexEsso** [comment]
i use this:
`var error;`
`final FirebaseUser user = (await auth
            .signInWithEmailAndPassword(
      email: _emailController.text,
      password: _passwordController.text,
    ).catchError((Object authError) {
              setState(() {
                error = authError;
              });
    })).user;`

and use it like this:
`Text('${error.message.toString()}'),`

**Comment by nikhilmufc7** [comment]
I made a gist with all the error codes listed in official flutter fire repository with examples
https://gist.github.com/nikhilmufc7/6b74a3c12a6e2d3284942d40ff583e37

Hopefully will be useful for others too

**Comment by TheBossProSniper** [comment]
Hey guys,

I have also recently faced this error and, I have found out that the ```.catchError()``` callback wasn't being called in the debug mode (which is when you click the ```Run->Start Debugging``` button in VSCode).

 However, when you type in flutter run -d <your simulator code>, then, the ```.catchError()``` method gets called back as it is not in debug mode. 

To get your preferred simulator's code paste this line of code in the terminal:
```
instruments -s devices
```
If that doesn't work, you can also try pasting this:
```
xcrun simctl list
```

Additionally, the app won't crash anymore with a ```PlatformException()``` and instead you will get a log like this one: 

```
[VERBOSE-2:ui_dart_state.cc(157)] Unhandled Exception: NoSuchMethodError: The getter 'uid' was called on null.
Receiver: null
```

I have been facing this problem in Google Sign In too, in which the ```.catchError()``` was not being called! 

In conclusion, if you have some error with handling errors in Firebase Authentication, you should first try to first run in through the terminal. Thanks, and I hope this helps!
